spec/toy_robot_spec.rb

```diff
@@ -31,48 +46,68 @@
     specify 'When y is more than four, it expect to raise an exception' do
       expect { subject.place(0,5,:north) }.to raise_error(InvalidPositionError, "Invalid Position, 0, 5, north")
     end
-    specify 'When valid it should move Robot to position and update facing' do
+    specify 'When given valid x and y, it should move Robot to position and update facing' do
       subject.place(3,4,:west)
       expect(subject.x).to eq(3)
       expect(subject.y).to eq(4)
       expect(subject.facing).to eq(:west)
     end
 
+    specify 'When given valid x and y, the array @movements should have only 1 item' do
+      subject.place(1,2,:north)
+      expect(subject.movements.length).to eq(1)
+    end
+
+    specify 'When given valid x and y, the 1st item in array should have the place coordinate and facing' do
+      subject.place(1,2,:north)
+      expect(subject.movements[0]).to eq("PLACE 1, 2, NORTH")
+    end
   end ## end of describe 'place'
 
   describe 'move' do
-    subject { ToyRobot.new }
     context 'When the Robot is at Origin (0, 0, :north)' do
-      it 'it will move one position forward' do
-        subject.move
-        expect(subject.x).to eq(0)
-        expect(subject.y).to eq(1)
-        expect(subject.facing).to eq(:north)
+      it 'will move one position forward' do
+        robot = ToyRobot.new
+        robot.move
+        expect(robot.x).to eq(0)
+        expect(robot.y).to eq(1)
+        expect(robot.facing).to eq(:north)
```

Comment:
Ahh... I see, so, when you write a test, just care about testing what is the actual purpose of the function.

Not the other 'possible' effects, so, for these other extra things just test it manually.

This is what worries me before, because I can not imagine how many tests you have to imagine to create to test a function.

I thought I have to spend sooo much time to write all kind of possibilities that might happen of the effect of a function.

Thus I would spending too much time unnecessarily.

Thank you for clarifying this... again, never learnt this lesson from the class...



________________________________
From: itinsley <notifications@github.com>
Sent: Friday, 26 August 2016 5:41 PM
To: KatherinaFatiguso/ToyRobotBasic
Cc: Katherina Fatiguso; State change
Subject: Re: [KatherinaFatiguso/ToyRobotBasic] Facing (#11)


In spec/toy_robot_spec.rb<https://github.com/KatherinaFatiguso/ToyRobotBasic/pull/11#discussion_r76376746>:

>      context 'When the Robot is at Origin (0, 0, :north)' do
> -      it 'it will move one position forward' do
> -        subject.move
> -        expect(subject.x).to eq(0)
> -        expect(subject.y).to eq(1)
> -        expect(subject.facing).to eq(:north)
> +      it 'will move one position forward' do
> +        robot = ToyRobot.new
> +        robot.move
> +        expect(robot.x).to eq(0)
> +        expect(robot.y).to eq(1)
> +        expect(robot.facing).to eq(:north)


try to avoid noise by re-testing for side effects that are unlikely. You cannot test everything. Move does not impact facing so don't re-assert it's value, moving north does not impact the X variable so don't re-assert it. Why not? Because all these extra tests, create noise that detracts from the clarity of the test. Look at my examples and they are much clearer. Set up, then test.
You do need to test for weird side effects but they should be done with manual testing - which should test the boundaries.
I think we can eventually refactor this to be much cleaner and test everything succinctly - but the journey is part of the process. That is the lesson.

-
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub<https://github.com/KatherinaFatiguso/ToyRobotBasic/pull/11/files/9d5b5e068848fe251edbf8e4183959059e5e209a#r76376746>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ALHdPQGNmtqHdOLT78dHng8uTjW_pwRgks5qjpi8gaJpZM4JspTg>.